File: frontend/src/api/context.ts
/**
 * Context API client — SKPL context management endpoints.
 *
 * Covers:
 * - Context generation & session summary
 * - Anatomy scanning & symbol search
 * - Bug logging & management
 * - Cerebrum memory
 * - Token tracking & waste detection
 */
import { client } from './client';

// ── Types ───────────────────────────────────────────────────────────────────

export interface ContextGenerationRequest {
  include_anatomy?: boolean;
  include_bugs?: boolean;
  include_memory?: boolean;
  max_anatomy_entries?: number;
  max_bug_entries?: number;
  max_memory_entries?: number;
}

export interface ContextGenerationResponse {
  context: string;
  estimated_tokens: number;
}

export interface SessionContextSummary {
  session_id: string;
  agent_id: string | null;
  created_at: string;
  anatomy: AnatomyStats | null;
  bugs: BugStats | null;
  memory: MemoryStats | null;
  tokens: TokenLedgerSummary | null;
  waste: WasteSummary | null;
  last_scan: ScanResultSummary | null;
}

export interface AnatomyStats {
  total_symbols: number;
  total_files: number;
  languages: Record<string, number>;
}

export interface BugStats {
  total: number;
  open: number;
  resolved: number;
  duplicate: number;
}

export interface MemoryStats {
  total_memories: number;
  by_category: Record<string, number>;
}

export interface TokenLedgerSummary {
  session_id: string;
  total_input_tokens: number;
  total_output_tokens: number;
  total_tokens: number;
  total_waste_tokens: number;
  waste_rate: number;
  total_cost_usd: number;
  entry_count: number;
  model_breakdown: Record<string, number>;
  provider_breakdown: Record<string, number>;
}

export interface WasteSummary {
  total_waste_tokens: number;
  waste_rate: number;
  pattern_count: number;
}

export interface ScanResultSummary {
  mode: string;
  files_scanned: number;
  symbols_extracted: number;
  duration_seconds: number;
}

// ── Scan ────────────────────────────────────────────────────────────────────

export interface ScanRequest {
  root_path?: string;
  mode?: 'full' | 'incremental';
  changed_files?: string[];
}

export interface ScanStatus {
  task_id: string;
  status: 'queued' | 'running' | 'completed' | 'failed';
  progress?: number;
  result?: ScanResultSummary;
  error?: string;
}

// ── Symbol ──────────────────────────────────────────────────────────────────

export interface SymbolEntry {
  id: string;
  name: string;
  kind: string;
  language: string;
  line_start: number;
  line_end: number;
  signature: string | null;
  description: string | null;
  parent: string | null;
  is_exported: boolean;
  file_path: string;
}

export interface SymbolSearchRequest {
  query: string;
  language?: string;
  kind?: string;
  limit?: number;
}

// ── Bug ─────────────────────────────────────────────────────────────────────

export interface BugEntry {
  id: string;
  session_id: string;
  agent_id: string | null;
  error_type: string;
  error_message: string;
  error_traceback: string | null;
  file_path: string | null;
  line_number: number | null;
  fingerprint: string | null;
  duplicate_of: string | null;
  status: string;
  resolution: string | null;
  resolved_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface LogBugRequest {
  error_type: string;
  error_message: string;
  error_traceback?: string;
  file_path?: string;
  line_number?: number;
}

export interface UpdateBugStatusRequest {
  status: string;
  resolution?: string;
}

// ── Memory ──────────────────────────────────────────────────────────────────

export interface MemoryEntry {
  id: string;
  key: string;
  value: string;
  category: string;
  confidence: number;
  ttl_seconds: number | null;
  access_count: number;
  last_accessed_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface RememberRequest {
  key: string;
  value: string;
  category?: string;
  confidence?: number;
}

// ── Waste ───────────────────────────────────────────────────────────────────

export interface WastePattern {
  pattern_type: string;
  severity: string;
  description: string;
  tokens_wasted: number;
  file_path: string | null;
  detected_at: string | null;
}

// ── API Functions ───────────────────────────────────────────────────────────

const BASE = '/contexts';

/** Generate a context string for a session. */
export function generateContext(
  sessionId: string,
  req: ContextGenerationRequest = {},
) {
  return client.post<ContextGenerationResponse>(
    `${BASE}/${sessionId}/generate`,
    req,
  );
}

/** Get a comprehensive session context summary. */
export function getSessionSummary(sessionId: string) {
  return client.get<SessionContextSummary>(`${BASE}/${sessionId}/summary`);
}

// ── Anatomy ─────────────────────────────────────────────────────────────────

/** Start an async anatomy scan. */
export function startScan(sessionId: string, req: ScanRequest = {}) {
  return client.post<ScanStatus>(`${BASE}/${sessionId}/anatomy/scan`, req);
}

/** Get scan task status. */
export function getScanStatus(sessionId: string, taskId: string) {
  return client.get<ScanStatus>(
    `${BASE}/${sessionId}/anatomy/scan/${taskId}`,
  );
}

/** Search anatomy symbols. */
export function searchSymbols(
  sessionId: string,
  req: SymbolSearchRequest,
) {
  return client.post<SymbolEntry[]>(
    `${BASE}/${sessionId}/anatomy/search`,
    req,
  );
}

/** Get anatomy store statistics. */
export function getAnatomyStats(sessionId: string) {
  return client.get<AnatomyStats>(`${BASE}/${sessionId}/anatomy/stats`);
}

// ── Bug Log ─────────────────────────────────────────────────────────────────

/** Log a bug. */
export function logBug(sessionId: string, req: LogBugRequest) {
  return client.post<BugEntry>(`${BASE}/${sessionId}/buglog`, req);
}

/** List recent bugs. */
export function listBugs(
  sessionId: string,
  limit?: number,
  status?: string,
) {
  const params: Record<string, string> = {};
  if (limit !== undefined) params.limit = String(limit);
  if (status) params.status = status;
  return client.get<BugEntry[]>(`${BASE}/${sessionId}/buglog`, params);
}

/** Update bug status. */
export function updateBugStatus(
  sessionId: string,
  bugId: string,
  req: UpdateBugStatusRequest,
) {
  return client.patch<BugEntry>(
    `${BASE}/${sessionId}/buglog/${bugId}`,
    req,
  );
}

/** Get bug statistics. */
export function getBugStats(sessionId: string) {
  return client.get<BugStats>(`${BASE}/${sessionId}/buglog/stats`);
}

// ── Cerebrum (Memory) ───────────────────────────────────────────────────────

/** Store a memory. */
export function remember(sessionId: string, req: RememberRequest) {
  return client.post<MemoryEntry>(`${BASE}/${sessionId}/cerebrum`, req);
}

/** List all memories for a session. */
export function listMemories(sessionId: string) {
  return client.get<MemoryEntry[]>(`${BASE}/${sessionId}/cerebrum`);
}

/** Recall a memory by key. */
export function recall(sessionId: string, key: string) {
  return client.get<MemoryEntry>(`${BASE}/${sessionId}/cerebrum/${key}`);
}

/** Update an existing memory. */
export function updateMemory(
  sessionId: string,
  key: string,
  req: RememberRequest,
) {
  return client.patch<MemoryEntry>(
    `${BASE}/${sessionId}/cerebrum/${key}`,
    req,
  );
}

/** Forget a memory. */
export function forget(sessionId: string, key: string) {
  return client.delete(`${BASE}/${sessionId}/cerebrum/${key}`);
}

/** Get memory statistics. */
export function getMemoryStats(sessionId: string) {
  return client.get<MemoryStats>(`${BASE}/${sessionId}/cerebrum/stats`);
}

// ── Memory System Health ────────────────────────────────────────────────────

export interface MemoryHealth {
  l1_cerebrum: boolean;
  l2_mem0: boolean;
  l3_knowledge: boolean;
}

/** Get memory system health status. */
export function getMemoryHealth() {
  return client.get<MemoryHealth>('/api/memory/health');
}

// ── Token Tracking ──────────────────────────────────────────────────────────

/** Get token usage summary. */
export function getTokenSummary(sessionId: string) {
  return client.get<TokenLedgerSummary>(`${BASE}/${sessionId}/tokens`);
}

/** Get detected waste patterns. */
export function getWastePatterns(sessionId: string) {
  return client.get<WastePattern[]>(`${BASE}/${sessionId}/tokens/waste`);
}

File: backend/src/skpl_agent/app/_router/_schema/_context.py
"""Request/response schemas for the context router.

Covers all context-related endpoints:
- Anatomy scanning
- Bug logging
- Memory (cerebrum)
- Token tracking
- Context generation
"""

from __future__ import annotations

from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field


# ---------------------------------------------------------------------------
# Anatomy Scanning
# ---------------------------------------------------------------------------


class ScanRequest(BaseModel):
    """Request to start an anatomy scan."""
    root_path: str = Field(
        default=".",
        description="Root directory to scan. Defaults to current directory.",
    )
    mode: str = Field(
        default="full",
        pattern="^(full|incremental)$",
        description="Scan mode: 'full' or 'incremental'.",
    )
    changed_files: list[str] = Field(
        default_factory=list,
        description="List of changed files for incremental scans.",
    )


class ScanStatusResponse(BaseModel):
    """Status of a scan task."""
    task_id: str = Field(description="Unique scan task identifier.")
    status: str = Field(description="Task status: queued, running, completed, failed, cancelled.")
    progress: int = Field(default=0, description="Number of files processed.")
    progress_total: int = Field(default=0, description="Total number of files to scan.")
    current_file: str = Field(default="", description="Currently processing file.")
    created_at: str | None = None
    started_at: str | None = None
    completed_at: str | None = None
    result: dict | None = Field(default=None, description="Scan result (only when completed).")
    error: str | None = Field(default=None, description="Error message (only when failed).")


class SymbolSearchRequest(BaseModel):
    """Request to search anatomy symbols."""
    query: str = Field(description="Search query for symbol names, signatures, or descriptions.")
    language: str | None = Field(default=None, description="Filter by programming language.")
    kind: str | None = Field(
        default=None,
        description="Filter by symbol kind: function, class, method, variable, interface, etc.",
    )
    limit: int = Field(default=50, ge=1, le=200, description="Max results to return.")


class SymbolResponse(BaseModel):
    """A single symbol from the anatomy store."""
    id: str
    name: str
    kind: str
    language: str
    line_start: int
    line_end: int
    signature: str | None = None
    description: str | None = None
    parent: str | None = None
    is_exported: bool = False
    file_path: str = ""


class AnatomyStatsResponse(BaseModel):
    """Anatomy store statistics."""
    total_symbols: int = 0
    total_files: int = 0
    languages: dict[str, int] = Field(default_factory=dict)
    backend: str = ""


# ---------------------------------------------------------------------------
# Bug Log
# ---------------------------------------------------------------------------


class LogBugRequest(BaseModel):
    """Request to log a bug."""
    error_type: str = Field(description="Error type, e.g., 'SyntaxError', 'ValueError'.")
    error_message: str = Field(description="Error message.")
    error_traceback: str | None = Field(default=None, description="Full traceback.")
    file_path: str | None = Field(default=None, description="File where the error occurred.")
    line_number: int | None = Field(default=None, description="Line number where the error occurred.")
    context_snippet: str | None = Field(default=None, description="Surrounding code context.")
    metadata: dict | None = Field(default=None, description="Additional metadata.")


class BugResponse(BaseModel):
    """A bug record."""
    id: str
    session_id: str
    agent_id: str | None = None
    error_type: str
    error_message: str
    error_traceback: str | None = None
    file_path: str | None = None
    line_number: int | None = None
    fingerprint: str
    duplicate_of: str | None = None
    status: str
    resolution: str | None = None
    resolved_at: str | None = None
    created_at: str
    updated_at: str


class UpdateBugStatusRequest(BaseModel):
    """Request to update bug status."""
    status: str = Field(description="New status: open, resolved, wont_fix, duplicate.")
    resolution: str | None = Field(default=None, description="Resolution description.")


class BugStatsResponse(BaseModel):
    """Bug statistics."""
    total: int = 0
    open: int = 0
    resolved: int = 0
    duplicates: int = 0
    by_type: dict[str, int] = Field(default_factory=dict)
    by_status: dict[str, int] = Field(default_factory=dict)


# ---------------------------------------------------------------------------
# Cerebrum (Memory)
# ---------------------------------------------------------------------------


class RememberRequest(BaseModel):
    """Request to store a memory."""
    key: str = Field(description="Memory key.")
    value: str = Field(description="Memory value.")
    category: str = Field(default="general", description="Memory category.")
    confidence: float = Field(default=1.0, ge=0.0, le=1.0, description="Confidence score (0-1).")
    ttl_seconds: int | None = Field(default=None, description="Time-to-live in seconds.")


class MemoryResponse(BaseModel):
    """A memory record."""
    id: str
    key: str
    value: str
    category: str
    confidence: float
    source: str | None = None
    ttl_seconds: int | None = None
    access_count: int = 0
    last_accessed_at: str | None = None
    created_at: str
    updated_at: str


class MemoryStatsResponse(BaseModel):
    """Memory statistics."""
    total_memories: int = 0
    by_category: dict[str, int] = Field(default_factory=dict)
    avg_confidence: float = 0.0
    total_accesses: int = 0


# ---------------------------------------------------------------------------
# Token Tracking
# ---------------------------------------------------------------------------


class TokenSummaryResponse(BaseModel):
    """Token usage summary."""
    session_id: str = ""
    total_input_tokens: int = 0
    total_output_tokens: int = 0
    total_tokens: int = 0
    total_waste_tokens: int = 0
    waste_rate: float = 0.0
    total_cost_usd: float = 0.0
    entry_count: int = 0
    model_breakdown: dict[str, int] = Field(default_factory=dict)
    provider_breakdown: dict[str, int] = Field(default_factory=dict)


class WastePatternResponse(BaseModel):
    """A detected waste pattern."""
    pattern_type: str
    severity: str
    description: str
    tokens_wasted: int
    file_path: str | None = None
    detected_at: str | None = None


# ---------------------------------------------------------------------------
# Context Generation
# ---------------------------------------------------------------------------


class ContextGenerationRequest(BaseModel):
    """Request to generate context string."""
    include_anatomy: bool = True
    include_bugs: bool = True
    include_memory: bool = True
    max_anatomy_entries: int = Field(default=50, ge=1, le=200)
    max_bug_entries: int = Field(default=10, ge=1, le=50)
    max_memory_entries: int = Field(default=50, ge=1, le=200)


class ContextGenerationResponse(BaseModel):
    """Generated context string."""
    context: str = Field(description="The generated context string.")
    estimated_tokens: int = Field(default=0, description="Estimated token count of the context.")


# ---------------------------------------------------------------------------
# Session Summary
# ---------------------------------------------------------------------------


class SessionContextSummaryResponse(BaseModel):
    """Comprehensive session context summary."""
    session_id: str
    agent_id: str | None = None
    created_at: str | None = None
    anatomy: AnatomyStatsResponse | None = None
    bugs: BugStatsResponse | None = None
    memory: MemoryStatsResponse | None = None
    tokens: TokenSummaryResponse | None = None
    waste: dict[str, int] | None = None